### PR TITLE
feat: Avoid cache poisoning on transcode fail

### PR DESCRIPTION
**Pull request by FoxBuru**
From the description on #1950, it seems the transcode part had two different problems to be tackled on:

- What to do if an error was detected on the async transcode cmd execution? By adding some way of notifying back the transcoder goroutine (by adding a channel to signal either EOF or file closure on the CachedStream instance), we can enforce a proper execution flow, according to StdoutPipe suggested ergonomics:

> Wait will close the pipe after seeing the command exit, so most callers need not close the pipe themselves. It is thus incorrect to call Wait before all reads from the pipe have completed.

Once we have this notification channel, we can put a request (via a callback to an invalidation to the cache instance) to purge a specific cache entry in relation to the current transcoding work, if cmd.Wait() returns an error by any reason. In that same reasoning, I've implemented a simple `Invalidate` method with some tests to provide guidance on what to expect when trying to invalidate a cache entry.

As a bonus, I've added a small check to assess if the `Content-Type` to be delivered matches to `audio/mp4`; and instead of the mime detected, try to return always `audio/aac` (Seems Safari is quite pungent on what mime types are streamable per se)

### CODE REVIEW

**Comment by deluan** [comment]
Hey, thanks for taking a look at this issue! I'll review the PR as soon as I can. In the meantime, please fix the build. Seems that you added a direct dependency to the go-multierror package, and you need to run `go mod tidy`.

**Comment by FoxBuru** [comment]
> Hey, thanks for taking a look at this issue! I'll review the PR as soon as I can. In the meantime, please fix the build. Seems that you added a direct dependency to the go-multierror package, and you need to run `go mod tidy`.

Sorry about that! Pushed an updated `go.mod` for that. Thanks!

**Comment by deluan** [comment]
Hey, I was looking into that as I also need to handle context cancelations (in case the http request is canceled). I'm having a hard time following the code. Regarding this comment from the docs:
> Wait will close the pipe after seeing the command exit, so most callers need not close the pipe themselves. It is thus incorrect to call Wait before all reads from the pipe have completed.

My understanding is that it incorrect to call `Wait` in this situation **IF** there's no other goroutine reading from it, or else it will cause a deadlock, waiting for the Pipe to close, but not reading from it.

Anyway, I tried a different approach, and implemented the Pipe with `io.Pipe()`, something like this:

```golang
func NewTranscodeJob(ctx context.Context, cmd string, args ...string) (*TranscodeJob, error) {
	a := []string{cmd}
	a = append(a, args...)
	j := &TranscodeJob{ctx: ctx, args: a}
	j.PipeReader, j.out = io.Pipe()
	err := j.start()
	go j.wait()
	if err != nil {
		return nil, err
	}
	return j, nil
}

type TranscodeJob struct {
	*io.PipeReader
	out  *io.PipeWriter
	ctx  context.Context
	args []string
	cmd *exec.Cmd
}

func (j *TranscodeJob) start() error {
	cmd := exec.CommandContext(j.ctx, j.args[0], j.args[1:]...) // #nosec
	cmd.Stdout = j.out
	cmd.Stderr = os.Stderr
	j.cmd = cmd

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("starting cmd: %w", err)
	}
	return nil
}

func (j *TranscodeJob) wait()  {
	var exitErr *exec.ExitError
	if err := j.cmd.Wait(); err != nil && !errors.As(err, &exitErr) {
		_ = j.out.CloseWithError(fmt.Errorf("waiting cmd: %w", err))
	}
	if code := j.cmd.ProcessState.ExitCode(); code > 1 {
		_ = j.out.CloseWithError(fmt.Errorf("%s exited with non-zero status code: %d", j.args[0], code))
	}
	if j.ctx.Err() != nil {
		_ = j.out.CloseWithError(j.ctx.Err())
	}
	_ = j.out.Close()
}
```

(I removed the call to `createTranscodeCommand()` to simplify the code)

I may be missing something, but this approach handle the following situations:
- command not found or failed to start: `NewTranscodeJob` will return error in this case
- command failed and return a non-zero exit code: The Reader (TranscodeJob) will return a error, causing the data to not be cached
- context is canceled: The Reader will return `context.Canceled` error, causing the data to not be cached

We would still need to be sure that errors returned by the Reader causes `fscache` to not cache the data (it should AFAIK)

What do you think of this approach?

File: server/subsonic/stream.go [hunk 1]
```diff
@@ -56,7 +56,13 @@ func (c *StreamController) Stream(w http.ResponseWriter, r *http.Request) (*resp
 	} else {
 		// If the stream doesn't provide a size (i.e. is not seekable), we can't support ranges/content-length
 		w.Header().Set("Accept-Ranges", "none")
-		w.Header().Set("Content-Type", stream.ContentType())
+		contentType := stream.ContentType()
+		switch contentType {
+		case "audio/mp4":
```
Comment (deluan) [inline server/subsonic/stream.go:61]: If we are hard coding this here, why not just set the mime type in the `consts/mime_types.go` file?

Anyway, is it safe to do this? What happens in other browsers?

**Comment by deluan** [comment]
Hey, I'm gonna close this in favour of the #2041 solution, as it has less moving parts. Thanks for the investigation and effort!

**Pull request closed without merging**